File: nodejs/src/ingestion/pipelines/sessionreplay/anonymize/url.ts
/**
 * URL scrub.
 *
 * - Numbers (a bare run of digits) are masked to `$` per digit (length-preserving; `$` rather than `#`
 *   so it doesn't clash with the fragment separator).
 * - Path: keep allow-listed segments; a number → `$$`; anything else → `[redacted]`.
 * - Query: a param survives only if its key or value is an allow-listed alphanumeric token. A number
 *   counts as denied for survival, so `id=42` is dropped while `page=2` → `page=$`. When a param
 *   survives, each side renders independently: allow-listed → kept, number → `$$`, denied → `[key]`/`[value]`.
 * - Fragment: kept only if it is an allow-listed alphanumeric token; numbers and everything else dropped.
 * - With `{ scrubAuthority: true }` it strips userinfo/port and
 *   collapses the host to `example.com` (keeping a leading allow-listed subdomain label).
 */
import { ScrubContext } from './config'
import { ScrubResult } from './text'

export interface UrlScrubOptions {
    scrubAuthority?: boolean
}

const SIMPLE = /^[A-Za-z0-9]*$/ // 100% alphanumeric (empty allowed)
const NUMERIC = /^[0-9]+$/ // a bare number

const maskNumber = (n: string): string => '$'.repeat(n.length) // `$`, not `#` (the fragment separator)

// An allow-listed, alphanumeric, non-number token — the only kind that keeps a query param/fragment
// alive.
function isAllowed(ctx: ScrubContext, t: string): boolean {
    return t.length > 0 && SIMPLE.test(t) && !NUMERIC.test(t) && ctx.allow.urlContains(t)
}

// Rendered form of a surviving token: a number → `$$`, an allow-listed token → itself, else null.
function renderToken(ctx: ScrubContext, t: string): string | null {
    return NUMERIC.test(t) ? maskNumber(t) : isAllowed(ctx, t) ? t : null
}

export function scrubUrl(ctx: ScrubContext, input: string, opts?: UrlScrubOptions): ScrubResult {
    // rrweb's standard blank-iframe placeholder: entropy-free, so redacting it only costs replay fidelity.
    if (input === 'about:blank') {
        return { value: input, changed: false }
    }
    const tailIdx = input.search(/[?#]/)
    const base = tailIdx === -1 ? input : input.slice(0, tailIdx)
    const tail = tailIdx === -1 ? '' : input.slice(tailIdx) // starts with ? or #
    let changed = false

    const { scheme, authority, path } = splitUrl(base)
    let out = scheme
    if (authority) {
        if (opts?.scrubAuthority) {
            const scrubbed = scrubHost(ctx, authority)
            if (scrubbed !== authority) {
                changed = true
            }
            out += scrubbed
        } else {
            out += authority
        }
    }

    let first = true
    for (const raw of path.split('/')) {
        if (first) {
            first = false
        } else {
            out += '/'
        }
        if (raw.length === 0) {
            continue
        }
        if (NUMERIC.test(raw)) {
            out += maskNumber(raw) // a bare number → $$
            changed = true
        } else if (isSafeSegment(raw) || ctx.allow.urlContains(raw)) {
            out += raw
        } else {
            out += '[redacted]'
            changed = true
        }
    }

    const tailOut = scrubTail(ctx, tail)
    if (tailOut !== tail) {
        changed = true
    }
    out += tailOut

    return { value: out, changed }
}

function scrubTail(ctx: ScrubContext, tail: string): string {
    if (tail === '') {
        return ''
    }
    let query = ''
    let frag = ''
    if (tail[0] === '?') {
        const h = tail.indexOf('#')
        query = h === -1 ? tail.slice(1) : tail.slice(1, h)
        frag = h === -1 ? '' : tail.slice(h + 1)
    } else {
        frag = tail.slice(1) // tail starts with '#'
    }

    let out = ''
    if (tail[0] === '?') {
        const kept: string[] = []
        for (const pair of query.split('&')) {
            if (pair === '') {
                continue
            }
            const eq = pair.indexOf('=')
            if (eq === -1) {
                if (isAllowed(ctx, pair)) {
                    kept.push(pair)
                }
                continue
            }
            const key = pair.slice(0, eq)
            const value = pair.slice(eq + 1)
            if (!isAllowed(ctx, key) && !(value !== '' && isAllowed(ctx, value))) {
                continue
            }
            const kr = renderToken(ctx, key) ?? '[key]'
            const vr = value === '' ? '' : (renderToken(ctx, value) ?? '[value]')
            kept.push(`${kr}=${vr}`)
        }
        if (kept.length > 0) {
            out += '?' + kept.join('&')
        }
    }
    if (frag.length > 0 && isAllowed(ctx, frag)) {
        out += '#' + frag
    }
    return out
}

// Strip userinfo + port and rewrite the host to example.com. Keep a leading *subdomain* label
// (only when there is one, i.e. ≥3 labels) if it's url-allow-listed: `us.test.com` → `us.example.com`.
function scrubHost(ctx: ScrubContext, authority: string): string {
    const at = authority.lastIndexOf('@')
    let host = at !== -1 ? authority.slice(at + 1) : authority
    host = host.replace(/:\d+$/, '') // drop :port
    const labels = host.split('.')
    const first = labels[0] ?? ''
    return labels.length > 2 && first && ctx.allow.urlContains(first) ? `${first}.example.com` : 'example.com'
}

// Split into scheme prefix (incl. `://` or `//`), authority (`[userinfo@]host[:port]`), and path.
function splitUrl(s: string): { scheme: string; authority: string; path: string } {
    let scheme = ''
    let rest = s
    const schemeEnd = s.indexOf('://')
    if (schemeEnd !== -1) {
        scheme = s.slice(0, schemeEnd + 3)
        rest = s.slice(schemeEnd + 3)
    } else if (s.startsWith('//')) {
        scheme = '//'
        rest = s.slice(2)
    } else {
        return { scheme: '', authority: '', path: s } // relative URL: all path
    }
    const pathOff = rest.indexOf('/')
    if (pathOff === -1) {
        return { scheme, authority: rest, path: '' }
    }
    return { scheme, authority: rest.slice(0, pathOff), path: rest.slice(pathOff) }
}

function isSafeSegment(seg: string): boolean {
    return seg === '' || seg === '.' || seg === '..'
}

File: rust/replay-anonymizer-node/src/mlhog/scrub/url.rs
// Copied from MLHog prep/labeling/src/scrub/url.rs — bench-only. Adapted: `crate::` import path
// points at `crate::mlhog::`.

//! URL scrub: keep allow-listed path/query/fragment tokens, mask numbers to `$`, redact the rest.
//! `scrub_authority` also strips userinfo/port and rewrites the host to example.com.

use crate::mlhog::context::Ctx;

#[allow(dead_code)]
pub fn scrub(ctx: &Ctx<'_>, input: &str) -> String {
    let mut out = String::with_capacity(input.len());
    scrub_into(ctx, input, &mut out);
    out
}

pub fn scrub_into(ctx: &Ctx<'_>, input: &str, out: &mut String) -> bool {
    scrub_into_opts(ctx, input, out, false)
}

pub fn scrub_into_authority(ctx: &Ctx<'_>, input: &str, out: &mut String) -> bool {
    scrub_into_opts(ctx, input, out, true)
}

fn scrub_into_opts(ctx: &Ctx<'_>, input: &str, out: &mut String, scrub_authority: bool) -> bool {
    let tail_idx = input.find(['?', '#']);
    let base = match tail_idx {
        Some(i) => &input[..i],
        None => input,
    };
    let tail = match tail_idx {
        Some(i) => &input[i..],
        None => "",
    };
    let mut changed = false;

    let (scheme, authority, path) = split_url(base);
    out.push_str(scheme);
    if !authority.is_empty() {
        if scrub_authority {
            let scrubbed = scrub_host(ctx, authority);
            if scrubbed != authority {
                changed = true;
            }
            out.push_str(&scrubbed);
        } else {
            out.push_str(authority);
        }
    }

    let mut first = true;
    for raw in path.split('/') {
        if first {
            first = false;
        } else {
            out.push('/');
        }
        if raw.is_empty() {
            continue;
        }
        if is_numeric(raw) {
            mask_number_into(raw, out);
            changed = true;
        } else if is_safe_segment(raw) || ctx.allow.url_contains(raw) {
            out.push_str(raw);
        } else {
            out.push_str("[redacted]");
            changed = true;
        }
    }

    let tail_out = scrub_tail(ctx, tail);
    if tail_out != tail {
        changed = true;
    }
    out.push_str(&tail_out);

    changed
}

fn scrub_tail(ctx: &Ctx<'_>, tail: &str) -> String {
    if tail.is_empty() {
        return String::new();
    }
    let (query, frag) = if tail.starts_with('?') {
        match tail.find('#') {
            Some(h) => (&tail[1..h], &tail[h + 1..]),
            None => (&tail[1..], ""),
        }
    } else {
        ("", &tail[1..])
    };

    let mut out = String::new();
    if tail.starts_with('?') {
        let mut kept: Vec<String> = Vec::new();
        for pair in query.split('&') {
            if pair.is_empty() {
                continue;
            }
            match pair.find('=') {
                None => {
                    if is_allowed(ctx, pair) {
                        kept.push(pair.to_string());
                    }
                }
                Some(eq) => {
                    let key = &pair[..eq];
                    let value = &pair[eq + 1..];
                    if !is_allowed(ctx, key) && !(!value.is_empty() && is_allowed(ctx, value)) {
                        continue;
                    }
                    let kr = render_token(ctx, key).unwrap_or_else(|| "[key]".to_string());
                    let vr = if value.is_empty() {
                        String::new()
                    } else {
                        render_token(ctx, value).unwrap_or_else(|| "[value]".to_string())
                    };
                    kept.push(format!("{kr}={vr}"));
                }
            }
        }
        if !kept.is_empty() {
            out.push('?');
            out.push_str(&kept.join("&"));
        }
    }
    if !frag.is_empty() && is_allowed(ctx, frag) {
        out.push('#');
        out.push_str(frag);
    }
    out
}

fn scrub_host(ctx: &Ctx<'_>, authority: &str) -> String {
    let mut host = match authority.rfind('@') {
        Some(at) => &authority[at + 1..],
        None => authority,
    };
    if let Some(colon) = host.rfind(':') {
        let port = &host[colon + 1..];
        if !port.is_empty() && port.bytes().all(|b| b.is_ascii_digit()) {
            host = &host[..colon];
        }
    }
    let labels: Vec<&str> = host.split('.').collect();
    let first = labels.first().copied().unwrap_or("");
    if labels.len() > 2 && !first.is_empty() && ctx.allow.url_contains(first) {
        format!("{first}.example.com")
    } else {
        "example.com".to_string()
    }
}

fn split_url(s: &str) -> (&str, &str, &str) {
    let (scheme, rest) = if let Some(scheme_end) = s.find("://") {
        (&s[..scheme_end + 3], &s[scheme_end + 3..])
    } else if let Some(rest) = s.strip_prefix("//") {
        (&s[..2], rest)
    } else {
        return ("", "", s);
    };
    match rest.find('/') {
        None => (scheme, rest, ""),
        Some(path_off) => (scheme, &rest[..path_off], &rest[path_off..]),
    }
}

fn is_simple(t: &str) -> bool {
    t.bytes().all(|b| b.is_ascii_alphanumeric())
}

fn is_numeric(t: &str) -> bool {
    !t.is_empty() && t.bytes().all(|b| b.is_ascii_digit())
}

fn mask_number_into(n: &str, out: &mut String) {
    for _ in 0..n.len() {
        out.push('$');
    }
}

fn mask_number(n: &str) -> String {
    "$".repeat(n.len())
}

fn is_allowed(ctx: &Ctx<'_>, t: &str) -> bool {
    !t.is_empty() && is_simple(t) && !is_numeric(t) && ctx.allow.url_contains(t)
}

fn render_token(ctx: &Ctx<'_>, t: &str) -> Option<String> {
    if is_numeric(t) {
        Some(mask_number(t))
    } else if is_allowed(ctx, t) {
        Some(t.to_string())
    } else {
        None
    }
}

fn is_safe_segment(seg: &str) -> bool {
    matches!(seg, "" | "." | "..")
}
